hypothesis-python/src/hypothesis/internal/conjecture/shrinking/dfas.py

```diff
@@ -0,0 +1,337 @@
+    of this function. If ``allowed_to_update`` is False, this will raise an
+    error as soon as it encounters a failure to normalize.
+
+    Additionally, if more than ``max_dfas` DFAs are required to normalize
+    this test function, this function will raise an error - it's essentially
+    designed for small patches that other shrink passes don't cover, and
+    if it's learning too many patches then you need a better shrink pass
+    than this can provide.
+    """
+    # Need import inside the function to avoid circular imports
+    from hypothesis.internal.conjecture.engine import BUFFER_SIZE, ConjectureRunner
+
+    runner = ConjectureRunner(
+        test_function,
+        settings=settings(database=None, suppress_health_check=HealthCheck.all()),
+        ignore_limits=True,
+    )
+
+    seen = set()
+
+    dfas_added = 0
+
+    found_interesting = False
+    consecutive_successes = 0
+    failures_to_find_interesting = 0
+    while consecutive_successes < required_successes:
+        attempt = runner.cached_test_function(b"", extend=BUFFER_SIZE)
+        if attempt.status < Status.INTERESTING:
+            failures_to_find_interesting += 1
+            assert (
+                found_interesting or failures_to_find_interesting <= 1000
+            ), "Test function seems to have no interesting test cases"
+            continue
+
+        found_interesting = True
+
+        target = attempt.interesting_origin
+
+        def shrinking_predicate(d):
+            return d.status == Status.INTERESTING and d.interesting_origin == target
+
+        if target not in seen:
+            seen.add(target)
+            runner.shrink(attempt, shrinking_predicate)
+            continue
+
+        previous = fully_shrink(
+            runner, runner.interesting_examples[target], shrinking_predicate
+        )
+        current = fully_shrink(runner, attempt, shrinking_predicate)
+
+        if current.buffer == previous.buffer:
+            consecutive_successes += 1
+            continue
+
+        consecutive_successes = 0
+
+        if not allowed_to_update:
+            raise FailedToNormalise(
+                "Shrinker failed to normalize %r to %r and we are not allowed to learn new DFAs."
+                % (previous.buffer, current.buffer)
+            )
+
+        if dfas_added >= max_dfas:
+            raise FailedToNormalise(
+                "Test function is too hard to learn: Added %d DFAs and still not done."
+                % (dfas_added,)
+            )
+
+        dfas_added += 1
+
+        new_dfa = learn_a_new_dfa(
+            runner, previous.buffer, current.buffer, shrinking_predicate
+        )
+
+        name = (
+            base_name
+            + "-"
+            + hashlib.sha256(repr(new_dfa).encode("utf-8")).hexdigest()[:10]
+        )
+
+        # If there is a name collision this DFA should already be being
+        # used for shrinking, so we should have current this case better
+        # then we already have.
```

Comment:
Is `current` here supposed to be something else? I can't make sense of this sentence as-is.

Reply:
Yeah, this sentence was nonsense, not sure what I was thinking. Replaced it with a better one.